### PR TITLE
redo(cdc): add a version field for redo

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #8168

### What is changed and how it works?

Add a version field into redo codec. So we can support more improvements later.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/model/codec/codec.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+)
+
+// NOTE: why we need this?
+//
+// Before this logic is introduced, redo log is encoded into byte slice without a version field.
+// This makes it hard to extend in future.
+// However in the old format (i.e. v1 format), the first 5 bytes are always same, which can be
+// confirmed in v1/codec_gen.go. So we reuse those bytes, and add a version field in them.
+var (
+	v1Header      [v1HeaderLength]byte      = [...]byte{0x83, 0xa3, 0x72, 0x6f}
```
Comment (CharlesCheung96) [inline cdc/model/codec/codec.go:40]:
```suggestion
	v1Header  = [v1HeaderLength]byte{0x83, 0xa3, 0x72, 0x6f}
```

File: cdc/model/codec/codec.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	versionPrefix [versionPrefixLength]byte = [...]byte{0xff, 0xff}
```
Comment (CharlesCheung96) [inline cdc/model/codec/codec.go:41]:
```suggestion
	versionPrefix = [versionPrefixLength]byte{0xff, 0xff}
```

File: cdc/model/codec/codec.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+			}
+			postUnmarshal(r)
+		} else {
+			panic("unsupported codec version")
+		}
+	}
+	return
+}
+
+func EncodeRedoLog(r *model.RedoLog, en *msgp.Writer) (err error) {
```
Comment (CharlesCheung96) [inline cdc/model/codec/codec.go:152]: Where is this method used？

File: cdc/model/codec/codec.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+			postUnmarshal(r)
+		} else {
+			panic("unsupported codec version")
+		}
+	}
+	return
+}
+
+func MarshalRedoLog(r *model.RedoLog, b []byte) (o []byte, err error) {
+	preMarshal(r)
```
Comment (CharlesCheung96) [inline cdc/model/codec/codec.go:201]: `preMarshal` was a little confused. Maybe we can encapsulate such logic as `ToRedoLog` method?

Ref: https://github.com/pingcap/tiflow/pull/8107/files#diff-eb40c4235fa377719c71b4d416eb38d4f4134b281758dd43687ca525d0ca3549R272

File: cdc/model/codec/codec.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+	versionPrefixLength int = 2
+	versionFieldLength  int = 2
+
+	latestVersion uint16 = 2
+)
+
+// NOTE: why we need this?
+//
+// Before this logic is introduced, redo log is encoded into byte slice without a version field.
+// This makes it hard to extend in future.
```
Comment (3AceShowHand) [inline cdc/model/codec/codec.go:36]:
```suggestion
// This makes it hard to extend in the future.
```

File: cdc/model/codec/codec.go (hunk 2 above)
Comment (3AceShowHand) [reply cdc/model/codec/codec.go:41]: where the header's value come from?

File: cdc/model/codec/codec.go (hunk 4 above)
Comment (hicqu) [reply cdc/model/codec/codec.go:201]: I don't want to expose interface `ToRedoLog`. All logic in the function is for marshaling. Maybe we can add a function `MarshalRowAsRedoLog`.

File: cdc/model/codec/codec.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+			}
+			r.RedoRow.PreColumns = append(r.RedoRow.PreColumns, redoC)
+		}
+	}
+	if r.RedoDDL.DDL != nil {
+		r.RedoDDL.Type = byte(r.RedoDDL.DDL.Type)
+	}
+}
+
+func DecodeRedoLog(dc *msgp.Reader) (r *model.RedoLog, err error) {
```
Comment (3AceShowHand) [inline cdc/model/codec/codec.go:111]: It looks this method is not used.

File: cdc/model/codec/codec.go (hunk 3 above)
Comment (3AceShowHand) [reply cdc/model/codec/codec.go:152]: Ditto

File: cdc/model/codec/codec.go (hunk 4 above)
Comment (hicqu) [reply cdc/model/codec/codec.go:150]: Done. PTAL again, thanks!

File: cdc/model/codec/codec.go (hunk 2 above)
Comment (hicqu) [reply cdc/model/codec/codec.go]: v1Header comes from v1/codec_gen.go. versionPrefix is defined by us.

**Review by 3AceShowHand — APPROVED** [review]

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**